// @doist/todoist-api-typescript/dist/utils/urlHelpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatDateToYYYYMMDD = formatDateToYYYYMMDD;
exports.getTaskUrl = getTaskUrl;
exports.getProjectUrl = getProjectUrl;
exports.getSectionUrl = getSectionUrl;
var endpoints_1 = require("../consts/endpoints");
/**
 * Formats a Date object to YYYY-MM-DD string format.
 *
 * @internal
 * @param date The Date object to format.
 * @returns The formatted date string in YYYY-MM-DD format.
 */
function formatDateToYYYYMMDD(date) {
    var year = date.getFullYear();
    var month = String(date.getMonth() + 1).padStart(2, '0');
    var day = String(date.getDate()).padStart(2, '0');
    return "".concat(year, "-").concat(month, "-").concat(day);
}
/**
 * Generate the URL for a given task.
 *
 * @param taskId The ID of the task.
 * @param content The content of the task.
 * @returns The URL string for the task view.
 */
function getTaskUrl(taskId, content) {
    var slug = content ? slugify(content) : undefined;
    var path = slug ? "".concat(slug, "-").concat(taskId) : taskId;
    return "".concat(endpoints_1.TODOIST_WEB_URI, "/task/").concat(path);
}
/**
 * Generate the URL for a given project.
 *
 * @param projectId The ID of the project.
 * @param name The name of the project.
 * @returns The URL string for the project view.
 */
function getProjectUrl(projectId, name) {
    var slug = name ? slugify(name) : undefined;
    var path = slug ? "".concat(slug, "-").concat(projectId) : projectId;
    return "".concat(endpoints_1.TODOIST_WEB_URI, "/project/").concat(path);
}
/**
 * Generate the URL for a given section.
 *
 * @param sectionId The ID of the section.
 * @param name The name of the section.
 * @returns The URL string for the section view.
 */
function getSectionUrl(sectionId, name) {
    var slug = name ? slugify(name) : undefined;
    var path = slug ? "".concat(slug, "-").concat(sectionId) : sectionId;
    return "".concat(endpoints_1.TODOIST_WEB_URI, "/section/").concat(path);
}
/**
 * Slugify function borrowed from Django.
 *
 * @param value The string to slugify.
 * @returns The slugified string.
 */
function slugify(value) {
    // Convert to ASCII
    var result = value.normalize('NFKD').replace(/[\u0300-\u036f]/g, '');
    // Remove non-ASCII characters
    result = result.replace(/[^\x20-\x7E]/g, '');
    // Convert to lowercase and replace non-alphanumeric characters with dashes
    result = result.toLowerCase().replace(/[^\w\s-]/g, '');
    // Replace spaces and repeated dashes with single dashes
    result = result.replace(/[-\s]+/g, '-');
    // Strip dashes from the beginning and end
    return result.replace(/^-+|-+$/g, '');
}
