// @doist/todoist-api-typescript/dist/consts/endpoints.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ENDPOINT_REVOKE_TOKEN = exports.ENDPOINT_GET_TOKEN = exports.ENDPOINT_AUTHORIZATION = exports.ENDPOINT_SYNC = exports.ENDPOINT_SYNC_QUICK_ADD = exports.PROJECT_UNARCHIVE = exports.PROJECT_ARCHIVE = exports.ENDPOINT_REST_ACTIVITIES = exports.ENDPOINT_REST_PRODUCTIVITY = exports.ENDPOINT_REST_USER = exports.ENDPOINT_REST_PROJECT_COLLABORATORS = exports.ENDPOINT_REST_PROJECTS_ARCHIVED = exports.ENDPOINT_REST_PROJECTS = exports.ENDPOINT_REST_TASK_REOPEN = exports.ENDPOINT_REST_TASK_CLOSE = exports.ENDPOINT_REST_COMMENTS = exports.ENDPOINT_REST_LABELS_SHARED_REMOVE = exports.ENDPOINT_REST_LABELS_SHARED_RENAME = exports.ENDPOINT_REST_LABELS_SHARED = exports.ENDPOINT_REST_LABELS = exports.ENDPOINT_REST_SECTIONS = exports.ENDPOINT_REST_TASKS_COMPLETED_SEARCH = exports.ENDPOINT_REST_TASKS_COMPLETED_BY_DUE_DATE = exports.ENDPOINT_REST_TASKS_COMPLETED_BY_COMPLETION_DATE = exports.ENDPOINT_REST_TASKS_FILTER = exports.ENDPOINT_REST_TASKS = exports.API_BASE_URI = exports.API_VERSION = exports.TODOIST_WEB_URI = void 0;
exports.getSyncBaseUri = getSyncBaseUri;
exports.getAuthBaseUri = getAuthBaseUri;
var BASE_URI = 'https://api.todoist.com';
var TODOIST_URI = 'https://todoist.com';
exports.TODOIST_WEB_URI = 'https://app.todoist.com/app';
// The API version is not configurable, to ensure
// compatibility between the API and the client.
exports.API_VERSION = 'v1';
exports.API_BASE_URI = "/api/".concat(exports.API_VERSION, "/");
var API_AUTHORIZATION_BASE_URI = '/oauth/';
function getSyncBaseUri(domainBase) {
    if (domainBase === void 0) { domainBase = BASE_URI; }
    return new URL(exports.API_BASE_URI, domainBase).toString();
}
function getAuthBaseUri(domainBase) {
    if (domainBase === void 0) { domainBase = TODOIST_URI; }
    return new URL(API_AUTHORIZATION_BASE_URI, domainBase).toString();
}
exports.ENDPOINT_REST_TASKS = 'tasks';
exports.ENDPOINT_REST_TASKS_FILTER = exports.ENDPOINT_REST_TASKS + '/filter';
exports.ENDPOINT_REST_TASKS_COMPLETED_BY_COMPLETION_DATE = exports.ENDPOINT_REST_TASKS + '/completed/by_completion_date';
exports.ENDPOINT_REST_TASKS_COMPLETED_BY_DUE_DATE = exports.ENDPOINT_REST_TASKS + '/completed/by_due_date';
exports.ENDPOINT_REST_TASKS_COMPLETED_SEARCH = 'completed/search';
exports.ENDPOINT_REST_SECTIONS = 'sections';
exports.ENDPOINT_REST_LABELS = 'labels';
exports.ENDPOINT_REST_LABELS_SHARED = exports.ENDPOINT_REST_LABELS + '/shared';
exports.ENDPOINT_REST_LABELS_SHARED_RENAME = exports.ENDPOINT_REST_LABELS_SHARED + '/rename';
exports.ENDPOINT_REST_LABELS_SHARED_REMOVE = exports.ENDPOINT_REST_LABELS_SHARED + '/remove';
exports.ENDPOINT_REST_COMMENTS = 'comments';
exports.ENDPOINT_REST_TASK_CLOSE = 'close';
exports.ENDPOINT_REST_TASK_REOPEN = 'reopen';
exports.ENDPOINT_REST_PROJECTS = 'projects';
exports.ENDPOINT_REST_PROJECTS_ARCHIVED = exports.ENDPOINT_REST_PROJECTS + '/archived';
exports.ENDPOINT_REST_PROJECT_COLLABORATORS = 'collaborators';
exports.ENDPOINT_REST_USER = 'user';
exports.ENDPOINT_REST_PRODUCTIVITY = exports.ENDPOINT_REST_TASKS + '/completed/stats';
exports.ENDPOINT_REST_ACTIVITIES = 'activities';
exports.PROJECT_ARCHIVE = 'archive';
exports.PROJECT_UNARCHIVE = 'unarchive';
exports.ENDPOINT_SYNC_QUICK_ADD = exports.ENDPOINT_REST_TASKS + '/quick';
exports.ENDPOINT_SYNC = 'sync';
exports.ENDPOINT_AUTHORIZATION = 'authorize';
exports.ENDPOINT_GET_TOKEN = 'access_token';
exports.ENDPOINT_REVOKE_TOKEN = 'access_tokens/revoke';
